File: algoritmos/orden/timsort/timsort.py
MIN_MERGE = 32
 
def calcMinRun(n): 
    r = 0
    while n >= MIN_MERGE: 
        r |= n & 1
        n >>= 1
    return n + r 

#funcion de insercion
def insertion(arr, left, right): 
    for i in range(left + 1, right + 1): 
        j = i 
        #elementos mayores que temp a la derecha
        while j > left and arr[j] < arr[j - 1]: 
            arr[j], arr[j - 1] = arr[j - 1], arr[j] 
            j -= 1

#funcion de mezcla
def merge(arr, l, m, r): 
    len1 = m - l + 1
    len2 = r - m 
    left = []
    right = [] 

    #copia datos a los arrays temporales
    for i in range(0, len1): 
        left.append(arr[l + i]) 
    for i in range(0, len2): 
        right.append(arr[m + 1 + i]) 

    i = 0
    j = 0
    k = l

    #combina los array
    while i < len1 and j < len2: 
        if left[i] <= right[j]: 
            arr[k] = left[i] 
            i += 1
        else: 
            arr[k] = right[j] 
            j += 1
        k += 1
    
    #copia elementos restantes de left
    while i < len1: 
        arr[k] = left[i] 
        k += 1
        i += 1
    
    #copia elementos restantes de right
    while j < len2: 
        arr[k] = right[j] 
        k += 1
        j += 1

#funcion principal del timSort
def timSort(arr): 
    n = len(arr) 
    minRun = calcMinRun(n) 
    #ordena los bloques de tamanio RUN con insertion
    for start in range(0, n, minRun): 
        end = min(start + minRun - 1, n - 1) 
        insertion(arr, start, end) 
    
    #combina los bloques ordenados
    size = minRun 
    while size < n: 

        for left in range(0, n, 2 * size): 
            mid = min(n - 1, left + size - 1) 
            right = min((left + 2 * size - 1), (n - 1)) 

            #combina solo si hay array para combinar
            if mid < right: 
                merge(arr, left, mid, right) 

        size = 2 * size 


def readRandomBin():
    try:
        with open("../random_list.bin", "rb") as file:
            array_bytes = file.read()
    except:
        print("Error al abrir")
        exit(1)

    num = len(array_bytes) // 4
    array_int = [int.from_bytes(array_bytes[i:i+4], byteorder='little') for i in range(0, len(array_bytes), 4)]

    return num, array_int

def createRandomFile(array_int):
    try:
        with open("random_list_py.txt", "w") as txt_file:
            for num_entero in array_int:
                txt_file.write(str(num_entero) + "\n")
    except:
        print("Error al crear el archivo de texto")
        exit(1)

def createSortedFile(array_int):
    try:
        with open("sorted_list_py.txt", "w") as txt_file_sorted:
            for num_entero in array_int:
                txt_file_sorted.write(str(num_entero) + "\n")
    except:
        print("Error al crear el archivo de texto")
        exit(1)


if __name__ == "__main__": 

    # archivo original
    num, array_int = readRandomBin()        
    
    """
    # txt para comprobar
    createRandomFile(array_int)
    """

    timSort(array_int)

    # txt para comprobar
    createSortedFile(array_int)

File: algoritmos/orden/timsort/timsort.c
#include <stdio.h>
#include <stdlib.h>

#define MIN_MERGE 32

int minRunLength(int n){
    int r = 0;
    while (n >= MIN_MERGE){
        r |= n & 1;
        n >>= 1;
    }
    return n + r;
}

// funcion  del insertion sort
void insertionSort(int arr[], int left, int right) {
    for (int i = left + 1; i <= right; i++) {
        int temp = arr[i];
        int j = i - 1;

        while (j >= left && arr[j] > temp) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }

        arr[j + 1] = temp;
    }
}

// funcion de merge
void merge(int arr[], int l, int m, int r) {
    int len1 = m - l + 1;
    int len2 = r - m;

    int *left = (int *)malloc(len1 * sizeof(int));
    int *right = (int *)malloc(len2 * sizeof(int));

    //copia datos a los arrays temporales
    for (int i = 0; i < len1; i++){
        left[i] = arr[l + i];
    }
        
    for (int j = 0; j < len2; j++){
        right[j] = arr[m + 1 + j];
    }

    int i = 0;
    int j = 0;
    int k = l;

    // combina los array
    while (i < len1 && j < len2) {
        if (left[i] <= right[j]) {
            arr[k] = left[i];
            i++;
        } else {
            arr[k] = right[j];
            j++;
        }
        k++;
    }

    // copia elementos restantes de left
    while (i < len1) {
        arr[k] = left[i];
        i++;
        k++;
    }

    // copia elementos restantes de right
    while (j < len2) {
        arr[k] = right[j];
        j++;
        k++;
    }

    free(left);
    free(right);
}

int min(int a, int b) {
    if (a < b)
        return a;
    else
        return b;
}

// funcion del timsort
void timSort(int arr[], int n) {
    int minRun = minRunLength(n);

    for (int i = 0; i < n; i += minRun){
        insertionSort(arr, i, min((i + minRun - 1), (n - 1)));
    }

    for (int size = minRun; size < n; size = 2 * size) {
        for (int left = 0; left < n; left += 2 * size) {
            int mid = left + size - 1;
            int right = (left + 2 * size - 1 < n - 1) ? (left + 2 * size - 1) : (n - 1);

            if (mid < right)
                merge(arr, left, mid, right);
        }
    }
}



int readRandomBin(int **array_int, int *num){
    FILE *file = fopen("../random_list.bin", "rb");
    if (file == NULL) {
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    *num = file_size / sizeof(int);
    *array_int = (int *)malloc(*num * sizeof(int));

    fread(*array_int, sizeof(int), *num, file);

    fclose(file);
    return 0;
}

int createRandomFile(int *array_int, int num){
    FILE *txt_file = fopen("random_list_c.txt", "w");
    if (txt_file == NULL) {
        return 1;
    }

    for (int i = 0; i < num; i++) {
        fprintf(txt_file, "%d\n", array_int[i]);
    }

    fclose(txt_file);
    return 0;
}

int createSortedFile(int *array_int, int num){
    FILE *txt_file_sorted = fopen("sorted_list_c.txt", "w");
    if (txt_file_sorted == NULL) {
        return 1;
    }

    for (int i = 0; i < num; i++) {
        fprintf(txt_file_sorted, "%d\n", array_int[i]);
    }

    fclose(txt_file_sorted);
    return 0;
}

int main() {
    int *array_int;
    int num;

    if (readRandomBin(&array_int, &num) != 0){
        printf("Error reading file\n");
        return 1;
    }

    /*
    if (createRandomFile(array_int, num) != 0){
        printf("Error creating file\n");
        return 1;
    }
    */

    timSort(array_int, num);

    if (createSortedFile(array_int, num) != 0){
        printf("Error creating file\n");
        return 1;
    }
    
    free(array_int);
    return 0;
}
